### ISSUE TITLE
No deadlock on macOS with POSIX threads

### ISSUE THREAD

**Issue by lemmy** [lemmy/BlockingQueue#3]
## TLDR: On *macOS*, don't lock a mutex that has not been initialized, don't wait on a condition that has not been initialized.

Everything that is written below is super interesting and plausible.  However, it is all wrong!  Spurious wakeups are a red herring, and the real problem is the way more mundane/trivial coding bug below that only manifests on macOS.
The missing initialization would have been detectable on macOS with a test that asserts the consistency of the elements that are enqueued and dequeued.  It is also obvious, that trace validation would have spotted the bug immediately.

```c
diff --git a/impl/producer_consumer.c b/impl/producer_consumer.c
index bf7efda..46cfdfe 100644
--- a/impl/producer_consumer.c
+++ b/impl/producer_consumer.c
@@ -73,6 +73,9 @@ int main(int argc, char * argv[]) {
 
        printf("Buffer size = %d, # Producers = %d, # Consumers = %d\n", buff_size, numProducers, numConsumers);
 
+       pthread_mutex_init(&mutex, NULL);
+       pthread_cond_init(&modify, NULL);
+
        /* Allocate space for the buffer */
        buffer = malloc(sizeof(int) * buff_size);
        pthread_t prods[numProducers], cons[numConsumers];
```

Works with:
```bash
markus@avocado:~$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
```
Fails with:
```bash
markus@armbook ~ % gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.27)
Target: arm64-apple-darwin20.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
-----

# Red herring: spurious wake-ups

The [C code](https://github.com/lemmy/BlockingQueue/blob/master/impl/producer_consumer.c) (POSIX) deadlocks on Windows and Linux but doesn't deadlocked on Intel (MacBook Pro Retina Early 2015) or Apple Silicon M1 mac.  However, the [Cocoa](https://en.wikipedia.org/wiki/Cocoa_(API)) variant [below](https://github.com/lemmy/BlockingQueue/issues/3#issuecomment-743940422) deadlocks as expected.  This seemingly has to do with the likelihood of `pthread_cond.c` causing [spurious wake-ups](https://en.wikipedia.org/wiki/Spurious_wakeup) (signal is effectively broadcast?) on macOS, which POSIX generally [allows](https://pubs.opengroup.org/onlinepubs/009604599/functions/pthread_cond_signal.html) for [performance reasons](https://stackoverflow.com/a/8594644).  Also see a related [SO discussion](https://stackoverflow.com/questions/8594591/why-does-pthread-cond-wait-have-spurious-wakeups).  

----

- 10.7.0 (Lion): http://www.opensource.apple.com/source/Libc/Libc-763.11/pthreads/pthread_cond.c
- 10.8.0 (Mountain Lion): http://www.opensource.apple.com/source/Libc/Libc-825.24/pthreads/pthread_cond.c
- 10.10.0 (Yosemite): http://www.opensource.apple.com/source/libpthread/libpthread-105.1.4/src/pthread_cond.c
- 10.11.0 (El Capitan): http://www.opensource.apple.com/source/libpthread/libpthread-137.1.1/src/pthread_cond.c
- 10.12.0 (Sierra): http://www.opensource.apple.com/source/libpthread/libpthread-218.1.3/src/pthread_cond.c
- 10.13.0 (High Sierra): http://www.opensource.apple.com/source/libpthread/libpthread-301.1.6/src/pthread_cond.c
- 10.14.0 (Mojave): http://www.opensource.apple.com/source/libpthread/libpthread-330.201.1/src/pthread_cond.c
- 10.14.1 (Mojave/latest): http://www.opensource.apple.com/source/libpthread/libpthread-330.220.2/src/pthread_cond.c

(list copied from a Rust issue at https://github.com/rust-lang/rust/issues/62291)

**Comment by lemmy** [comment]
Minimum configuration (1,2,1) doesn't deadlock on an Apple M1 (ARM) either [thanks to a volunteer on Twitter](https://twitter.com/lemmster/status/1337477984991625219).

**Comment by lemmy** [comment]
Deadlocks on FreeBSD running on an Azure Standard B2s.

```bash
markus@freebsd:~ $ ./producer_consumer 2 3 7
Buffer size = 2, # Producers = 3, # Consumers = 7

1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
^C
markus@freebsd:~ $ ./producer_consumer 2 3 99
Buffer size = 2, # Producers = 3, # Consumers = 99

1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
1 values consumed
^C
markus@freebsd:~ $ ./producer_consumer 2 1 99
Buffer size = 2, # Producers = 1, # Consumers = 99

1 values consumed
1 values consumed
1 values consumed
1 values consumed
^C
markus@freebsd:~ $ uname -a
FreeBSD freebsd 12.2-RELEASE-p1 FreeBSD 12.2-RELEASE-p1 GENERIC  amd64
```

**Comment by lemmy** [comment]
(Setup)
```bash
markus@armbook impl % git log -n1
commit 74eeccdf98badde40c865033d6bb81aa27d06814 (HEAD, tag: v03_FSM)
Author: Markus Alexander Kuppe <tlaplus.net@lemmster.de>
Date:   Thu Jan 16 21:15:57 2020 -0800

    v03 (FSM): State graph for configuration p1c2b1.
    
    Slightly different configuration with which we can visually spot the
    deadlock.

markus@armbook impl % pwd
/Users/markus/blockingqueue/impl
```

Silicon Mac (arm64):

```bash
markus@armbook impl % uname -ma
Darwin armbook 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:10 PDT 2020; root:xnu-7195.50.7~2/RELEASE_ARM64_T8101 arm64
markus@armbook impl % gcc -target arm64-apple-macos11 -o producer_consumer producer_consumer.c
markus@armbook impl % lipo -archs producer_consumer
arm64
markus@armbook impl % ./producer_consumer 2 1 99
# ...it runs and runs and runs...
```

Silicon Mac (x86_64/rosetta 2):

```bash
markus@armbook impl % uname -ma
Darwin armbook 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:10 PDT 2020; root:xnu-7195.50.7~2/RELEASE_ARM64_T8101 arm64
markus@armbook impl % make
markus@macbook ~ % lipo -archs producer_consumer
x86_64
markus@armbook impl % ./producer_consumer 2 1 99
# ...it runs and runs and runs...
```

Intel Mac:

```bash
markus@macbook ~ % uname -ma
Darwin macbook 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:11 PDT 2020; root:xnu-7195.50.7~2/RELEASE_X86_64 x86_64
markus@macbook ~ % lipo -archs producer_consumer
x86_64
markus@armbook impl % ./producer_consumer 2 1 99
# ...it runs and runs and runs...
```

**Comment by lemmy** [comment]
Printing consumers ids doesn't show anything unusual.  The schedule runs all consumers:

`arm64`
```bash
markus@armbook impl % head -50 arm64.txt                                            
Buffer size = 1, # Producers = 1, # Consumers = 99

1 values consumed by 0
1 values consumed by 1
1 values consumed by 2
10001 values consumed by 0
10001 values consumed by 1
10001 values consumed by 2
20001 values consumed by 0
20001 values consumed by 1
20001 values consumed by 2
30001 values consumed by 0
30001 values consumed by 1
30001 values consumed by 2
40001 values consumed by 0
40001 values consumed by 1
40001 values consumed by 2
1 values consumed by 3
1 values consumed by 5
1 values consumed by 4
1 values consumed by 6
1 values consumed by 7
1 values consumed by 8
50001 values consumed by 0
50001 values consumed by 1
1 values consumed by 9
1 values consumed by 10
50001 values consumed by 2
10001 values consumed by 5
60001 values consumed by 0
60001 values consumed by 1
60001 values consumed by 2
20001 values consumed by 5
1 values consumed by 11
10001 values consumed by 8
30001 values consumed by 5
10001 values consumed by 9
10001 values consumed by 4
1 values consumed by 12
10001 values consumed by 6
20001 values consumed by 8
1 values consumed by 13
40001 values consumed by 5
20001 values consumed by 9
1 values consumed by 14
1 values consumed by 15
1 values consumed by 16
20001 values consumed by 6
30001 values consumed by 8
```

`x86_64`
```bash
markus@armbook impl % head -50 x86.txt 
Buffer size = 1, # Producers = 1, # Consumers = 99

1 values consumed by 3
1 values consumed by 2
1 values consumed by 4
1 values consumed by 1
1 values consumed by 6
1 values consumed by 44
1 values consumed by 45
1 values consumed by 46
1 values consumed by 90
10001 values consumed by 1
10001 values consumed by 6
10001 values consumed by 44
10001 values consumed by 90
10001 values consumed by 46
20001 values consumed by 1
10001 values consumed by 45
20001 values consumed by 6
30001 values consumed by 1
30001 values consumed by 6
1 values consumed by 7
10001 values consumed by 2
10001 values consumed by 3
10001 values consumed by 7
20001 values consumed by 2
20001 values consumed by 3
1 values consumed by 8
1 values consumed by 5
1 values consumed by 0
20001 values consumed by 7
1 values consumed by 9
10001 values consumed by 4
10001 values consumed by 5
30001 values consumed by 3
10001 values consumed by 0
30001 values consumed by 7
10001 values consumed by 9
20001 values consumed by 5
20001 values consumed by 0
10001 values consumed by 8
20001 values consumed by 9
40001 values consumed by 7
30001 values consumed by 5
30001 values consumed by 0
20001 values consumed by 4
30001 values consumed by 9
50001 values consumed by 7
1 values consumed by 10
40001 values consumed by 5
```

```c
markus@armbook impl % git diff
diff --git a/impl/producer_consumer.c b/impl/producer_consumer.c
index bf7efda..26e60a4 100644
--- a/impl/producer_consumer.c
+++ b/impl/producer_consumer.c
@@ -41,6 +41,7 @@ void *producer (void * arg) {
 }
 
 void *consumer (void * arg) {
+       uint32_t id = *((uint32_t *) arg);
        long report = 0;
        while(1) {
                pthread_mutex_lock(&mutex);   // acquire the lock
@@ -53,7 +54,7 @@ void *consumer (void * arg) {
                pthread_mutex_unlock(&mutex); // release the lock
                
                if (report++ % 10000 == 0) {
-               printf("\n%ld values consumed", report);
+               printf("\n%ld values consumed by %d", report, id);
                }
        } 
 }
@@ -76,6 +77,7 @@ int main(int argc, char * argv[]) {
        /* Allocate space for the buffer */
        buffer = malloc(sizeof(int) * buff_size);
        pthread_t prods[numProducers], cons[numConsumers];
+       uint32_t threadIds[numConsumers];
 
        uint32_t i;
        /* Create the producer */
@@ -84,8 +86,10 @@ int main(int argc, char * argv[]) {
        }
 
        /* Create the consumers */
-       for (i = 0; i < numConsumers; i++)
-               pthread_create(&cons[i], NULL, consumer, NULL);
+       for (i = 0; i < numConsumers; i++) {
+               threadIds[i] = i;
+               pthread_create(&cons[i], NULL, consumer, &threadIds[i]);
+       }
 
        /* Wait for all threads to finish */
        for (i = 0; i < numProducers ; i++) 
```

**Comment by lemmy** [comment]
Contrary to the [pthread variant](https://github.com/lemmy/BlockingQueue/blob/master/impl/producer_consumer.c), this deadlocks immediately on macOS (for the expected configurations such as 1 2 1).


`clang -framework Foundation producer_consumer_cocoa.m -o producer_consumer_cocoa`

```objective-c
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 4) {
            printf ("Usage: ./producer_consumer <buffer_size> <#_of_producers> <#_of_consumers> \n");
            exit(1);
        }

        /* Process arguments */
        const int buff_size = atoi(argv[1]);
        const int numProducers = atoi(argv[2]);
        const int numConsumers = atoi(argv[3]);

        printf("Buffer size = %d, # Producers = %d, # Consumers = %d\n", buff_size, numProducers, numConsumers);
        
        // This lock just keeps the main thread from terminating
        // but has nothing to do with the deadlock we want to
        // study.
        NSLock *waitForever = [NSLock new];

        // This is the interesting lock (https://developer.apple.com/documentation/foundation/nscondition)
        NSCondition *lock = [NSCondition new];
        
        // Simplified queue to a simple counter.
        __block int count = 0;
    
        for (int i = 0; i < numProducers; i++) {
            [NSThread detachNewThreadWithBlock:^{
                while (YES) {
                    [lock lock];
                    while (count == buff_size) {
                        [lock wait];
                    }
                    count = count + 1;
                    // Replace signal with broadcast here and
                    // below to workaround the deadlock.
                    [lock signal];
                    [lock unlock];
                }
            }];
        }

        for (int i = 0; i < numConsumers; i++) {
            [NSThread detachNewThreadWithBlock:^{
                long report = 0;
                while (YES) {
                    [lock lock];
                    while (count == 0) {
                        [lock wait];
                    }
                    count = count - 1;
                    [lock signal];
                    [lock unlock];
                    if (report++ % 10000 == 0) {
                        printf("\n%ld values consumed by %d", report, i);
                    }
                }
            }];
        }

        [waitForever lock];
        [waitForever lock]; // wait forever until users kills the process.
    }
    return 0;
}
```

**Comment by kylog** [comment]
Oh boy, total red herring. Uninitialized variables strike again. Nice debugging! :)

**Comment by lemmy** [comment]
https://github.com/lemmy/BlockingQueue/commit/4ca68e4e4d7e9fd246ec49f4cf3916ecc81fe15c#diff-9b1f603b300eeb3dd333cd5f8f308278fea7c951428fa8d972a630c01468c4d6R78-R79

**Comment by lemmy** [comment]
Background reading: 

https://code.woboq.org/userspace/glibc/nptl/pthread_cond_signal.c.html
https://probablydance.com/2020/10/31/using-tla-in-the-real-world-to-understand-a-glibc-bug/
